Utilities/TestFatDxe/TestFatDxe.c

```diff
@@ -0,0 +1,504 @@
+/** @file
+  Copyright (c) 2023, Savva Mitrofanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+**/
+
+#include <Fat.h>
+
+#include <UserFile.h>
+#include <UserGlobalVar.h>
+#include <UserMemory.h>
+#include <UserUnicodeCollation.h>
+#include <string.h>
+
+extern EFI_UNICODE_COLLATION_PROTOCOL  *mUnicodeCollationInterface;
+
+STATIC EFI_SIMPLE_FILE_SYSTEM_PROTOCOL  *mEfiSfsInterface;
+
+UINTN        mFuzzOffset;
+UINTN        mFuzzSize;
+CONST UINT8  *mFuzzPointer;
+
+EFI_STATUS
+EFIAPI
+FuzzBlockIoFlushBlocks (
+  IN EFI_BLOCK_IO_PROTOCOL  *This
+  )
+{
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+FuzzReadDisk (
+  IN  EFI_DISK_IO_PROTOCOL  *This,
+  IN  UINT32                MediaId,
+  IN  UINT64                Offset,
+  IN  UINTN                 BufferSize,
+  OUT VOID                  *Buffer
+  )
+{
+  if (Buffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((mFuzzSize - mFuzzOffset) < BufferSize) {
+    return EFI_DEVICE_ERROR;
+  }
+
+  CopyMem (Buffer, mFuzzPointer, BufferSize);
+  mFuzzPointer += BufferSize;
+  mFuzzOffset  += BufferSize;
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+WrapInstallMultipleProtocolInterfaces (
+  IN OUT EFI_HANDLE  *Handle,
+  ...
+  )
+{
+  VA_LIST     Args;
+  EFI_STATUS  Status;
+  EFI_GUID    *Protocol;
+  VOID        *Interface;
+
+  if (Handle == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  VA_START (Args, Handle);
+  for (Status = EFI_SUCCESS; !EFI_ERROR (Status);) {
+    //
+    // If protocol is NULL, then it's the end of the list
+    //
+    Protocol = VA_ARG (Args, EFI_GUID *);
+    if (Protocol == NULL) {
+      break;
+    }
+
+    Interface = VA_ARG (Args, VOID *);
+
+    //
+    // If this is Sfs protocol then save interface into global state
+    //
+    if (CompareGuid (Protocol, &gEfiSimpleFileSystemProtocolGuid)) {
+      mEfiSfsInterface = Interface;
+    }
+  }
+
+  VA_END (Args);
+
+  return Status;
+}
+
+VOID
+FreeAll (
+  IN CHAR16             *FileName,
+  IN FAT_VOLUME         *Volume,
+  IN EFI_FILE_PROTOCOL  *RootDirInstance
+  )
+{
+  FreePool (FileName);
+
+  //
+  // Closes root directory
+  //
+  if (RootDirInstance != NULL) {
+    FatClose (RootDirInstance);
+  }
+
+  //
+  // Abandon volume structure
+  //
+  if (Volume != NULL) {
```

Comment:
This does not look right to me. Why does not Fat itself do it and we need to manually free the internals like DiskIo or BlockIo, which we did not allocate?

Reply:
Well, we definitely allocate both `BlockIo` and `DiskIo` [here](https://github.com/acidanthera/OpenCorePkg/blob/d55bf707e9bfccaa11713a48824533ce4ef6db86/Utilities/TestFatDxe/TestFatDxe.c#L235-L264) and then call `FatAllocateVolume`, [which prepares the Volume structure.](https://github.com/acidanthera/audk/blob/18722f129ff1f06347daa0504906903ffa785bcc/FatPkg/EnhancedFatDxe/Init.c#L50-L54) This is the same scheme we used in TestNtfsDxe ([1](https://github.com/acidanthera/OpenCorePkg/blob/savvas-fatpkg-fuzz-20230301/Utilities/TestNtfsDxe/TestNtfsDxe.c#L128-L152), [2](https://github.com/acidanthera/OpenCorePkg/blob/savvas-fatpkg-fuzz-20230301/Utilities/TestNtfsDxe/TestNtfsDxe.c#L50-L60)) and TestExt4Dxe ([1](https://github.com/acidanthera/OpenCorePkg/blob/savvas-fatpkg-fuzz-20230301/Utilities/TestExt4Dxe/TestExt4Dxe.c#L279-L306), [2](https://github.com/acidanthera/OpenCorePkg/blob/savvas-fatpkg-fuzz-20230301/Utilities/TestExt4Dxe/TestExt4Dxe.c#L146-L156)). 

The EnhancedFatDxe itself allocates these interfaces inside DriverBindingStart using [`OpenProtocol`](https://github.com/acidanthera/audk/blob/master/FatPkg/EnhancedFatDxe/Fat.c#L384-L422) call, which support we don't implement in User library. And deallocates, at least DiskIo, in DriverBindingStop using [`CloseProtocol`](https://github.com/acidanthera/audk/blob/master/FatPkg/EnhancedFatDxe/Fat.c#L516-L532). Maybe if we implement such logic in the User library, then we can make protocol usage look more native than self-allocate such structures. But it's not so easy.